hw/drivers/sensors/bme680/src/bme680.c

```diff
@@ -0,0 +1,1694 @@
+{
+	int8_t rslt = BME680_OK;
+
+	if (value != NULL) {
+		/* Check if value is below minimum value */
+		if (*value < min) {
+			/* Auto correct the invalid value to minimum value */
+			*value = min;
+			dev->info_msg |= BME680_I_MIN_CORRECTION;
+		}
+		/* Check if value is above maximum value */
+		if (*value > max) {
+			/* Auto correct the invalid value to maximum value */
+			*value = max;
+			dev->info_msg |= BME680_I_MAX_CORRECTION;
+		}
+	} else {
+		rslt = BME680_E_NULL_PTR;
+	}
+
+	return rslt;
+}
+
+/*!
+ * @brief This internal API is used to validate the device structure pointer for
+ * null conditions.
+ */
+static int8_t null_ptr_check(const struct bme680_cfg *dev)
+{
+	int8_t rslt;
+
+	if ((dev == NULL) || (dev->read == NULL) || (dev->write == NULL) || (dev->delay_ms == NULL)) {
+		/* Device structure pointer is not valid */
+		rslt = BME680_E_NULL_PTR;
+	} else {
+		/* Device structure is fine */
+		rslt = BME680_OK;
+	}
+
+	return rslt;
+}
+
+static void
+bme680_delay_ms(uint32_t period)
+{
+	uint32_t ticks;
+
+	os_time_ms_to_ticks(period, &ticks);
+	os_time_delay(ticks);
+}
+
+static int8_t
+bme680_spi_read(struct sensor *sensor, uint8_t dev_id, uint8_t reg_addr, uint8_t *reg_data, uint16_t len)
+{
+	struct sensor_itf *interface = SENSOR_GET_ITF(sensor);
+
+	hal_gpio_write(interface->si_cs_pin, 0);
+	hal_spi_tx_val(interface->si_num, reg_addr | 0x80);
+	/* Use reg_data for the txbuf as well. The contents don't matter since we're reading */
+	hal_spi_txrx(interface->si_num, reg_data, reg_data, len);
+	hal_gpio_write(interface->si_cs_pin, 1);
+
+	return 0;
+}
+
+static int8_t
+bme680_spi_write(struct sensor *sensor, uint8_t dev_id, uint8_t reg_addr, uint8_t *reg_data, uint16_t len)
+{
+	struct sensor_itf *interface = SENSOR_GET_ITF(sensor);
+
+	hal_gpio_write(interface->si_cs_pin, 0);
+	hal_spi_tx_val(interface->si_num, reg_addr);
+	hal_spi_txrx(interface->si_num, reg_data, NULL, len);
+	hal_gpio_write(interface->si_cs_pin, 1);
+
+	return 0;
+}
+
+static int8_t
+bme680_i2c_read(struct sensor *sensor, uint8_t dev_id, uint8_t reg_addr, uint8_t *reg_data, uint16_t len)
+{
+	int8_t rc = 0;
+	struct sensor_itf *interface = SENSOR_GET_ITF(sensor);
+	struct hal_i2c_master_data data;
+	data.address = dev_id;
+
+	/* Send the register to read from */
+	data.buffer = &reg_addr;
+	data.len = 1;
+	rc = hal_i2c_master_write(interface->si_num, &data, OS_TICKS_PER_SEC, 0);
+	if (rc != 0) {
+		goto exit;
+	}
+
+	/* Now read the data */
+	data.buffer = reg_data;
+	data.len = len;
+	rc = hal_i2c_master_read(interface->si_num, &data, OS_TICKS_PER_SEC, 1);
+	if (rc != 0) {
+		goto exit;
+	}
+
+exit: 
+	return rc;
+}
+
+static int8_t
+bme680_i2c_write(struct sensor *sensor, uint8_t dev_id, uint8_t reg_addr, uint8_t *reg_data, uint16_t len)
+{
+	int8_t rc = 0;
+	struct sensor_itf *interface = SENSOR_GET_ITF(sensor);
+	struct hal_i2c_master_data data;
+	uint8_t *buf;
+
+	buf = malloc(len + 1);
```

Comment:
avoid malloc() and free() in running code -- how big are these lengths generally?  better: use a buffer on the stack if not too big, or malloc() a single buffer at init time.